data/src/main/java/me/androidbox/data/worker_manager/AgendaSynchronizerWorker.kt

```diff
@@ -0,0 +1,166 @@
+    @Assisted private val context: Context,
+    @Assisted private val workerParameters: WorkerParameters,
+    private val taskService: TaskService,
+    private val taskDao: TaskDao,
+    private val reminderService: ReminderService,
+    private val reminderDao: ReminderDao,
+) : CoroutineWorker(context, workerParameters) {
+
+    override suspend fun doWork(): Result {
+
+        /** TODO is this the correct place to have this check and to retry or fail */
+        if(runAttemptCount > RETRY_COUNT) {
+            Result.failure()
+        }
+        else {
+            Result.retry()
+        }
+
+        val createdTasks: Deferred<List<TaskEntity>>
+        val updatedTasks: Deferred<List<TaskEntity>>
+        val createdReminders: Deferred<List<ReminderEntity>>
+        val updatedReminders: Deferred<List<ReminderEntity>>
+
+        /** Get all the created, updated tasks and reminders
+         *  TODO As these will run sequentially - is it better to use supervisor scope and run them in parallel? */
+        supervisorScope {
+            createdTasks = async {
+                val taskIds = taskDao.getAllTasksBySyncType(SyncAgendaType.CREATE)
+
+                taskIds.map { id ->
+                    taskDao.getTaskById(id)
+                }
+            }
+
+            updatedTasks = async {
+                val taskIds = taskDao.getAllTasksBySyncType(SyncAgendaType.UPDATE)
+
+                taskIds.map { id ->
+                    taskDao.getTaskById(id)
+                }
+            }
+
+            createdReminders = async {
+                val reminderIds = reminderDao.getAllRemindersBySyncType(SyncAgendaType.CREATE)
+
+                reminderIds.map { id ->
+                    reminderDao.getReminderById(id)
+                }
+            }
+
+            updatedReminders = async {
+                val reminderIds = reminderDao.getAllRemindersBySyncType(SyncAgendaType.UPDATE)
+
+                reminderIds.map { id ->
+                    reminderDao.getReminderById(id)
+                }
+            }
+        }
+
+        /** TODO I am using supervisor scope to launch so if one fails the other will still run
+         *  However, will the worker return failure if a single one fails even though some failed?
+         *  My checkResult will run everything in this block */
+        val responseState = checkResult {
+            supervisorScope {
+                val createTaskJob = launch {
+                    createdTasks.await().forEach { taskEntity ->
+                        /* Create the task on remote */
+                        val result = checkResult {
+                            taskService.createTask(taskEntity.toTaskDto())
+                        }
+                        if(result.isSuccess) {
+                            /* If success will run this delete - else if loop next next one */
+                            taskDao.deleteSyncTaskById(taskEntity.id)
```

Comment:
Currently you don't consider the following scenario:
1. The user is in offline mode and adds a task
2. Then, while still being in offline mode, they delete the created task again
3. You'll end up with one entry for a created task and one entry with one deleted task, where both have the same task ID
4. This call will delete both, even though it only refers to deleting the created task. Therefore, the task that was actually deleted locally will not be deleted remotely since the DB entry is missing.

You can either handle this by also passing the sync type here to always only delete one entry or you already make sure to delete all other sync types of a task if it's deleted (which you might be doing, but at a different place - then don't worry)